receiver/scraperhelper/settings.go

```diff
@@ -29,26 +29,41 @@ type ScraperControllerSettings struct {
 	// InitialDelay sets the initial start delay for the scraper,
 	// any non positive value is assumed to be immediately.
 	InitialDelay time.Duration `mapstructure:"initial_delay"`
-	// Timeout is used to set scraper's context deadline, it must be within
-	// the range of (0, CollectionInterval]
+	// Timeout is an optional value used to set scraper's context deadline,
+	// it must be within the range of (0, CollectionInterval].
```

Comment:
It's not clear to me why we would constrain the timeout based on the collection interval. I would like to be able to set the timeout to something greater than the collection interval, so that a scrape may take a reasonable amount of additional time if necessary, but is killed in extreme cases. (e.g. 1.2x collection interval is fine, but 10x is not).

Reply:
I do think we need a bounded cap that is based collection interval on the timeout should be, however, since we don't have a means of surfacing how long a scrape took, I am happy to leave this unbounded until we can surface that data.

Reply:
> What's the rationale for needing a bounded cap [?]

I would like to improve the scheduling guarantees for scrapers since we currently don't offer any safeguards.

The concerns I have where scrapers run outside the collection interval:
- Miss-aligned data resolution between configuration and observability system
  - Could be configured to run every 10s but finishes every 30s
- Resource contention created by scraper
  - Some scrapers may invoke a _stop the world_ or a global lock process which could impact the actual function of the service.
- No delays between scrapes
  - In the event that each scrape takes as long as the collection interval to observe a resource, then the observed system's resources are always used by the collector which may have negative side effects.

Now, these are mostly theoretical scenarios that I can think of a bounded execution cap set that would help solve but I'd like to actually get the data to help validate the frequency of these cases happening. 

I am happy to keep it as an opt-in addition since I don't know the likelihood of these happening at this point.

> ... as opposed to letting users decide?

Sorry, not against letting users decide at all as I know this will more than likely be a power user feature.

Let me create an issue to discuss this further to best circulate this around.

Reply:
I'm happy to discuss on another issue but to clarify my question:

Assuming a user opts-in to a timeout, why shouldn't they be allowed to chose a timeout that is longer than the collection interval? I think this is a valid choice, (and the only way I would ever use it or recommend using it.)

Maybe I missed the reason here, but none of these seem to address the point.

> * Miss-aligned data resolution between configuration and observability system
>   
>   * Could be configured to run every 10s but finishes every 30s

This guarantees when we _start trying_ to collect data, but generally will have no positive effect on the result of collecting data. Consider two cases:

Case 1: `median time required to collect: 30s`, `collection_interval: 10s`, `timeout: 10s`

In this scenario the timeout will end up killing the collection attempt pretty much every time. The user will receive no data. The problem isn't solved but is instead exacerbated.

Case 2: `median time required to collect: 5s`, `collection_interval: 10s`, `timeout: 10s`

Collection will complete within the collection interval nearly every time. However, if it _does_ run long, we till end up with irregularities in the data reported. The timeout can help give a better chance that we _only_ miss one interval, but doesn't give us any guarantees either way.

> * Resource contention created by scraper
>   
>   * Some scrapers may invoke a _stop the world_ or a global lock process which could impact the actual function of the service.

I agree this would be worth killing, but it seems like a bug that needs to be addressed in the scraper. Is there any scraper where this is an actual possibility?

> * No delays between scrapes
>   
>   * In the event that each scrape takes as long as the collection interval to observe a resource, then the observed system's resources are always used by the collector which may have negative side effects.

A timeout that is less than the collection interval would help in this case, but I don't see that it has any bearing on why the timeout _may not_ be greater than the collection interval. Any timeout equal to or greater than the collection interval will result in no improvement in this case, so why allow the "equal" case but not the "greater than".

Reply:
Right, I see what you mean now. 

> This guarantees when we start trying to collect data, but generally will have no positive effect on the result of collecting data. 

I was working off the premise that this would happen infrequently but would be consistent intervals that are missed, for example, the docker socket is overwhelmed and needs to have resources freed up so it can continue. In this scenario is better to exit early and come back to retry. However, your point remains true since these intervals would report no values.

> I agree this would be worth killing, but it seems like a bug that needs to be addressed in the scraper. Is there any scraper where this is an actual possibility?

I don't know for certain, But I think the JMX, SQL, and expvar under certain conditions.

> A timeout that is less than the collection interval would help in this case, but I don't see that it has any bearing on why the timeout may not be greater than the collection interval. Any timeout equal to or greater than the collection interval will result in no improvement in this case, so why allow the "equal" case but not the "greater than".

I agree it won't fix the data guarantees that I'd like but I'd like to add timeout support (that has no restrictions on duration besides can't be neg) to help provide some controls around scraper execution.

Reply:
Thanks for working through the details with me. I think a timeout will be a great feature as long as it's optional and not constrained by the collection interval.